rabbitmq/src/main/java/io/micronaut/rabbitmq/intercept/RabbitMQConsumerAdvice.java

```diff
@@ -193,111 +201,213 @@ public void process(BeanDefinition<?> beanDefinition, ExecutableMethod<?, ?> met
+
+                DefaultConsumer consumer = createDefaultConsumer(method,
+                        consumerParams,
+                        reQueue,
+                        hasAckArg,
+                        channel,
+                        isVoid,
+                        bean,
+                        binder,
+                        executorService,
+                        arguments);
+                channel.basicConsume(queue, false, clientTag, false, exclusive, arguments, consumer);
+            } catch (Throwable e) {
+                if (!channel.isOpen()) {
+                    channelPool.returnChannel(channel);
+                    consumerChannels.remove(channel);
+                    if (LOG.isErrorEnabled()) {
+                        LOG.error("The channel was closed due to an exception. The consumer [{}] will no longer receive messages", clientTag);
+                    }
+                }
+                handleException(new RabbitListenerException("An error occurred subscribing to a queue", e, bean, null));
+            }
+        }
 
-                    @Override
-                    public void handleTerminate(String consumerTag) {
-                        if (channel instanceof RecoverableChannel) {
-                            if (LOG.isDebugEnabled()) {
-                                LOG.debug("The channel was been terminated.  Automatic recovery attempt is underway for consumer [{}]", clientTag);
-                            }
-                        } else {
-                            ConsumerState state = consumerChannels.remove(channel);
-                            if (state != null) {
-                                state.channelPool.returnChannel(channel);
-                                if (LOG.isDebugEnabled()) {
-                                    LOG.debug("The channel was terminated. The consumer [{}] will no longer receive messages", clientTag);
-                                }
-                            }
+    }
+
+    private DefaultConsumer createDefaultConsumer(ExecutableMethod<?, ?> method, ConsumerParams consumerParams, boolean reQueue, boolean hasAckArg, Channel channel, boolean isVoid, Object bean, DefaultExecutableBinder<RabbitConsumerState> binder, ExecutorService executorService, Map<String, Object> arguments) {
+        DefaultConsumer consumer = new DefaultConsumer() {
+            @Override
+            public void handleTerminate(String consumerTag) {
+                if (channel instanceof RecoverableChannel) {
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug(
+                                "The channel was terminated.  Automatic recovery attempt is underway for consumer [{}]",
+                                consumerParams.clientTag);
+                    }
+                    try {
+                        channel.basicConsume(consumerParams.queue, false, consumerParams.clientTag, false, consumerParams.exclusive, arguments, this);
```

Comment:
So in general it seems this PR is working around or trying to amend the functionality built into the java client API. How can you safely call consume here when its possible for RabbitMQ to auto recover? Couldn't this end up in multiple subscriptions?

Reply:
When RabbitMQ closes a channel, it is [due to an error](https://www.rabbitmq.com/channels.html#error-handling). RabbitMQ Java client explicitly [does not recover](https://www.rabbitmq.com/api-guide.html#recovery) from such failures. IMO no client should. This is a clear case where the error should be visible to the user.